### PR TITLE
Update cudnn/onednn helpers, minor refactoring changes for imports

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Updates cudnn helpers, minor refactoring changes for classes.
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ ] Created tests for any significant new code additions.
- [ ] Relevant tests for your changes are passing.

### CODE REVIEW

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/dropout/Dropout.java [hunk 1]
```diff
@@ -108,14 +110,38 @@ protected void initializeHelper(DataType dataType){
...
         if("CUDA".equalsIgnoreCase(backend)) {
-            helper = DL4JClassLoading.createNewInstance(
-                    "org.deeplearning4j.cuda.dropout.CudnnDropoutHelper",
-                    DropoutHelper.class,
-                    dataType);
-            log.debug("CudnnDropoutHelper successfully initialized");
-            if (!helper.checkSupported()) {
+            if(DL4JClassLoading.loadClassByName(CUDNN_DROPOUT_HELPER_CLASS_NAME) != null) {
+                helper = DL4JClassLoading.createNewInstance(
+                        "org.deeplearning4j.cuda.dropout.CudnnDropoutHelper",
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/dropout/Dropout.java]: You've already introduced a constant for the class name, might as well use it here too.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/dropout/Dropout.java [hunk 2]
```diff
@@ -108,14 +110,38 @@ protected void initializeHelper(DataType dataType){
...
+                        DropoutHelper.class,
+                        dataType);
+            } else {
+                log.warn("Cudnn class not found using current class loader. Trying current classloader.");
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/dropout/Dropout.java]: This message doesn't make any sense.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/ConvolutionLayer.java [hunk 3]
```diff
@@ -73,14 +73,39 @@ public ConvolutionLayer(NeuralNetConfiguration conf, DataType dataType) {
...
-                    dataType);
-            log.debug("CudnnConvolutionHelper successfully initialized");
-            if (!helper.checkSupported()) {
+            if(DL4JClassLoading.loadClassByName(CUDA_CNN_HELPER_CLASS_NAME) != null) {
+                helper = DL4JClassLoading.createNewInstance(
+                        CUDA_CNN_HELPER_CLASS_NAME,
+                        ConvolutionHelper.class,
+                        dataType);
+            } else {
+                log.warn("Cudnn class not found using current class loader. Trying current classloader.");
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/ConvolutionLayer.java]: Again this message doesn't make any sense.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/subsampling/SubsamplingLayer.java [hunk 4]
```diff
@@ -50,26 +51,54 @@
...
+            if(DL4JClassLoading.loadClassByName(CUDNN_SUBSAMPLING_HELPER_CLASS_NAME) != null) {
+                helper = DL4JClassLoading.createNewInstance(
+                        CUDNN_SUBSAMPLING_HELPER_CLASS_NAME,
+                        SubsamplingHelper.class,
+                        dataType);
+                log.debug("CudnnSubsamplingHelper successfully initialized");
+
+            }
+            else {
+                log.warn("Cudnn class not found using current class loader. Trying current classloader.");
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/subsampling/SubsamplingLayer.java]: And a third time that this message doesn't make any sense.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/BatchNormalization.java [hunk 5]
```diff
@@ -72,11 +73,33 @@ void initializeHelper() {
...
-            log.debug("CudnnBatchNormalizationHelper successfully initialized");
+            if(DL4JClassLoading.loadClassByName(BATCH_NORM_CUDNN_HELPER_CLASS_NAME) != null) {
+                helper = DL4JClassLoading.createNewInstance(
+                        BATCH_NORM_CUDNN_HELPER_CLASS_NAME,
+                        BatchNormalizationHelper.class,
+                        dataType);
+                log.debug("CudnnBatchNormalizationHelper successfully initialized");
+
+            } else {
+                log.warn("Cudnn class not found using current class loader. Trying current classloader.");
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/BatchNormalization.java]: And a fourth time... Had I known that this thing repeats so often, I wouldn't have flagged every single appearance of it.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/LocalResponseNormalization.java [hunk 6]
```diff
@@ -65,11 +65,35 @@ public LocalResponseNormalization(NeuralNetConfiguration conf, DataType dataType
...
-                    LocalResponseNormalizationHelper.class,
-                    dataType);
-            log.debug("CudnnLocalResponseNormalizationHelper successfully initialized");
+            if(DL4JClassLoading.loadClassByName(LOCAL_RESPONSE_NORM_CUDNN_HELPER_CLASS_NAME) != null) {
+                helper = DL4JClassLoading.createNewInstance(
+                        LOCAL_RESPONSE_NORM_CUDNN_HELPER_CLASS_NAME,
+                        LocalResponseNormalizationHelper.class,
+                        dataType);
+            } else {
+                log.warn("Cudnn class not found using current class loader. Trying current classloader.");
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/LocalResponseNormalization.java]: Fifth time....

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/recurrent/LSTM.java [hunk 7]
```diff
@@ -53,32 +54,52 @@ public LSTM(NeuralNetConfiguration conf, DataType dataType) {
...
-                    dataType);
-            log.debug("CudnnLSTMHelper successfully initialized");
-            if (!helper.checkSupported(layerConf().getGateActivationFn(), layerConf().getActivationFn(), false)) {
+            if(DL4JClassLoading.loadClassByName(CUDNN_LSTM_CLASS_NAME) != null) {
+                helper = DL4JClassLoading.createNewInstance(
+                        CUDNN_LSTM_CLASS_NAME,
+                        LSTMHelper.class,
+                        dataType);
+            } else {
+                log.warn("Cudnn class not found using current class loader. Trying current classloader.");
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/recurrent/LSTM.java]: 6th time....

**Review by treo — CHANGES_REQUESTED** [review]
Next time try to have whitespace-ish changes in a separate PR :)

As for the actual changes, as you see in all of the comments, I'm not a big fan of the message:
"Cudnn class not found using current class loader. Trying current classloader."


Overall, I think it wouldn't be too hard to replace this entire loading pattern with a static helper, as you now have 6 repetitions of the exactly same pattern all over the code.

**Comment by agibsonccc** [comment]
@treo  let me think about your comments regarding the static helper. Overall I agree. Do you mind looking at the current attempt for the error message with respect to each class? Maybe we can create a helper utils here.  Do you think we should put that in the DL4jClassLoading util itself or maybe in just the helper package? We might have to do something similar for mkldnn/onednn as well?

**Comment by treo** [comment]
> Do you mind looking at the current attempt for the error message with respect to each class?
> "[...] Trying to use class loader that loaded this class instead"
Depending on the logger configuration, this may be a rather meaningless statement. I'd probably explicitly put the class name into it. 

Also, please use the interpolation support of the logging library instead of using string concatenation here. 

> Maybe we can create a helper utils here. Do you think we should put that in the DL4jClassLoading util itself or maybe in just the helper package?

Most of the big similar blocks are related to class loading, so it makes sense to just move it there entirely. Some of the dissimilar pieces we might want to pull out into a lambda, but how exactly that works out we'll have to see when we actually try it.

> We might have to do something similar for mkldnn/onednn as well?
I haven't looked at the code related to it, but if it is similar to this, then probably yes.

**Comment by agibsonccc** [comment]
@treo eliminated the code repetition where I could. Mind looking again?

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                                                         String layerName) {
+        String backend = Nd4j.getExecutioner().getEnvironmentInformation().getProperty("backend");
+        T helperRet = null;
+        if("CUDA".equalsIgnoreCase(backend)) {
+            if(DL4JClassLoading.loadClassByName(cudnnHelperClassName) != null) {
+                helperRet =  DL4JClassLoading.createNewInstance(
+                        cudnnHelperClassName,
+                        layerHelperSuperClazz,
+                        dataType);
+                log.debug("Cudnn heper {} successfully initialized",cudnnHelperClassName);
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java]: Typo. Should be "helper"

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+     * @param dataType the datatype to be used in the layer
+     * @param layerHelperSuperClazz the layer helper super class
+     * @param layerName the name of the layer to be created
+     * @param <T> the actual class type to be returned
+     * @return
+     */
+    public static <T extends LayerHelper> T createHelper(String cudnnHelperClassName,
+                                                         String oneDnnClassName,
+                                                         DataType dataType,
+                                                         Class<? extends LayerHelper> layerHelperSuperClazz,
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java]: Why the `clazz` naming?

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            if(DL4JClassLoading.loadClassByName(cudnnHelperClassName) != null) {
+                helperRet =  DL4JClassLoading.createNewInstance(
+                        cudnnHelperClassName,
+                        layerHelperSuperClazz,
+                        dataType);
+                log.debug("Cudnn heper {} successfully initialized",cudnnHelperClassName);
+
+            }
+            else {
+                log.warn("Unable to find class {}  using the classloader set for Dl4jClassLoading. Trying to use class loader that loaded this class instead.",cudnnHelperClassName);
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java]: You can't say "this class" in this context anymore. The logger will point to helper utils, not the class that is loading the helper.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/dropout/Dropout.java [hunk 11]
```diff
@@ -108,14 +110,38 @@ protected void initializeHelper(DataType dataType){
...
+                }
+            }
+
+            if (helper != null && !helper.checkSupported()) {
                 helper = null;
             }
+
+            if(helper != null) {
+                log.debug("CudnnDropoutHelper successfully initialized");
+            }
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/dropout/Dropout.java]: You're not using your new HelperUtils here.

**Review by treo — CHANGES_REQUESTED** [review]

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java (hunk 9 above)
Comment (agibsonccc) [reply deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java]: Habit from spring days. Changed.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java (hunk 10 above)
Comment (agibsonccc) [reply deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java]: Fixed.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            if(DL4JClassLoading.loadClassByName(cudnnHelperClassName) != null) {
+                helperRet =  DL4JClassLoading.createNewInstance(
+                        cudnnHelperClassName,
+                        layerHelperSuperClass,
+                        dataType);
+                log.debug("Cudnn helper {} successfully initialized",cudnnHelperClassName);
+
+            }
+            else {
+                log.warn("Unable to find class {}  using the classloader set for Dl4jClassLoading. Trying to use class loader that loaded the  class {} instead.",cudnnHelperClassName,cudnnHelperClassName);
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java]: This still doesn't make any sense. I think you meant to use `..., cudnnHelperClassName,layerHelperSuperClass.getCanonicalName());` or something like that?

**Review by treo — CHANGES_REQUESTED** [review]
Almost there, the log message is still not done yet.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java (hunk 12 above)
Comment (treo) [reply deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/HelperUtils.java]: This still hasn't been addressed?

**Review by treo — CHANGES_REQUESTED** [review]
The comment about the log message hasn't been addressed.

**Review by treo — APPROVED** [review]

**Pull request merged**